hubble/utils/jwt_parser.py

```diff
@@ -0,0 +1,44 @@
+import base64
+import json
+
+from jose import jwt
+
+from .jwks import JSONWebKeySet
+
+
+def decode_jwt(token: str):
+    """Decode component of JWT
+
+    :param token: base64 encoded JWT component (header or payload)
+    """
+
+    decoded = base64.b64decode(token).decode('utf-8')
+    try:
+        return json.loads(decoded)
+    except Exception:
+        return decoded
+
+
+def validate_jwt(token: str):
+    """Decode and validate JWT signature against Hubble public key
+
+    :param token: jwt string (as received from Hubble)
+    """
+    components = token.split('.')
+    # NOTE: this is needed because of some base64 decoding errors
+    header = decode_jwt(components[0] + "========")
+
+    if header['alg'] != 'ES256':
+        raise Exception(f"Algorithm not supported {header['alg']}")
+
+    keys = JSONWebKeySet.get_keys(header['kid'])
+    if len(keys) <= 0:
+        raise Exception(f"Signing key not found {header['kid']}")
+
+    decoded = jwt.decode(
+        token,
+        json.dumps(keys[0]),
+        algorithms=["ES256"],
```

Comment:
The key/signing algorithm should not be pinned on the client side.
Is it possible to simply pass the keys to the underlying library and not specify algorithms?

On the server side, the signing algorithm is decided based on hostname, certificate, and identityAudience settings.
Where the certificate algorithm may be different per hostname and may be changed via helm chart and k8s configurations.
It will probably also not be `keys[0]` when there are multiple keys.

Reply:
The library actually recommends pinning the algorithms on the decoding side. It the case you described, we can simply provide more signing algorithms.

Reply:
Auth0 says it's okay to do though https://auth0.com/blog/how-to-handle-jwt-in-python/

Reply:
> It will probably also not be keys[0] when there are multiple keys.
We fetch the keys from config by `kid`, will there be multiple keys with the same `kid` @nomagick

Reply:
>We fetch the keys from config by kid, will there be multiple keys with the same kid

Well in that case, it should be OK.

For algorithms, I think we only need to filter out "none".
https://www.rfc-editor.org/rfc/rfc8725.html#section-2.1
Otherwise we'll simply list all valid asymmetrical algorithms here.

Reply:
What are the all supported algorithms? `"HS256", "RS256", "ES256"`?

Reply:
As of now, they are: 
<img width="381" alt="CleanShot 2023-01-31 at 21 35 17@2x" src="https://user-images.githubusercontent.com/565869/215774920-7433c57a-83a0-49fd-8aa6-13fedbc6f36c.png">

Where the recommended ones are "RS256", "ES256"

Reply:
HS* are filtered out intentionally. We only need asymmetrical algorithms here.